### DISCUSSION TITLE
Trivy k8s - private registry support for vulnerabilities scan

### DISCUSSION THREAD

**Discussion by Anc0r** [aquasecurity/trivy#4499] [Ideas]
### Description

Hi everyone!

We are using Trivy in several GitLab CI/CD pipelines.

Currently we are testing the possibility to perform a kuberetes scan. I noticed that, in contrast to scanning images, none of our images from the private registry could be scanned. While looking through the documentation of 'trivy k8s' I noticed that unlike 'trivy image' the parameters '--username' and '--password' are not present. (and therefore also not the environment variables)

So it seems that you can't give 'trivy k8s' the credentials for private registrations? Is that right? Could you, like with 'trivy image', add these parameters?

What I have tested
- 'trivy k8s' with TRIVY_PASSWORD & TRIVY_USERNAME
- 'trivy k8s' with configured trivy.yaml

### Target

Kubernetes

### Scanner

Vulnerability

**Comment by AnaisUrlichs** [comment]
Hi there, the usage/use cases for both `trivy image` and `trivy k8s` are different. Trivy image needs acess to remote private registries where images are stored. Similarly `trivy repo` for remote repository scans needs access to also private registries.
In comparison, `trivy k8s` uses the kubeconfig for the running Kubernetes cluster that you would like to scan to authenticate and access the cluster. You can either scan the cluster locally by having connected to it as the default kubeconfig or you can set the kubeconfig and then scan the cluster or you can pass in the kubeconfig to trivy when performing the scan:
```
trivy k8s --kubeconfig ~/.kube/config2 -n kube-system --report=summary all
```
https://aquasecurity.github.io/trivy/v0.41/docs/target/kubernetes/#commands

If you want to scan container images inside your cluster for vulnerabilities specifically, I would recommend scanning them with the `trivy image` command or using the [Trivy operator](https://github.com/aquasecurity/trivy-operator) installed inside your cluster for more continuous K8s scanning.

The trivy k8s command is mainly useful to go through cluster security issues incl. vulnerabilities (I am thinking here of different workflows but maybe other people perceive it differently).

**Reply by AnaisUrlichs** [reply]
Oh I might have misread your question, you want to be able to allow trivy to access the container images from private registries that are running inside your cluster to scan them for vulnerabilities? 

If you don't want to scan them before deployment or in addition scan them after deployment, you can configure that in the trivy operator much easier https://aquasecurity.github.io/trivy-operator/v0.13.2/tutorials/private-registries/

**Reply by Anc0r** [reply]
Exactly. 
We have container with images from our private registry.
For 'trivy image' I can simply give the credentials via '--username' and '--password' (respectively the environment variables).

This would be desirable for 'trivy k8s'.

After some trial and error I found out that I can enable access to the private registry by modifying the ' ~/.docker/config.json'. 
But as a pipeline solution to scan different clusters/namespaces, I find this unnecessarily complicated.

**Reply by Anc0r** [reply]
Unfortunately, the Trivy Operator is not so suitable for our use case.

**Reply by knqyf263** [reply]
@chen-keinan Can you please take a look?

**Reply by AnaisUrlichs** [reply]
hmm, so I tried to replicate something similar for my own understanding 
I pushed a container image to ghcr (private) and installed it in my cluster through a K8s deployment:
```
spec:
      containers:
      - name: cns-website
        image: ghcr.io/cloud-native-security/secret-repo:0.0.9
        ports:
 .....
       imagePullSecrets:
      - name: dockerconfigjson-github-com
```

with a pullsecret:
```
kind: Secret
type: kubernetes.io/dockerconfigjson
apiVersion: v1
metadata:
  name: dockerconfigjson-github-com
  namespace: demo
  labels:
    run: cns-website
data:
  .dockerconfigjson: <info for the ghcr credentials>
```

And trivy can scan that container image (I deleted it locally before I performed the cluster scan) so if trivy is not using the local credentials from docker but only the in-cluster info, then the scan of private images should work @chen-keinan 
<img width="1031" alt="Screenshot 2023-05-31 at 14 02 33" src="https://github.com/aquasecurity/trivy/assets/33576047/37879cdb-c6d6-4205-bec2-3a244222e80d">

**Reply by knqyf263** [reply]
@AnaisUrlichs You logged in `ghcr.io` with `docker login`, right? It works as mentioned above, but @Anc0r wants to pass credential via CLI.

>After some trial and error I found out that I can enable access to the private registry by modifying the ' ~/.docker/config.json'.

**Reply by AnaisUrlichs** [reply]
true, never mind  
<img width="1107" alt="Screenshot 2023-05-31 at 15 19 54" src="https://github.com/aquasecurity/trivy/assets/33576047/92ba76e2-488b-4120-b29b-705afeaff805">

**Comment by chen-keinan** [comment]
I have created issue #4517 for it
